pkg/reconciler/knativeserving/knativeserving_controller.go

```diff
@@ -246,11 +246,16 @@ func (r *Reconciler) delete(instance *servingv1alpha1.KnativeServing) error {
 	if len(instance.GetFinalizers()) == 0 || instance.GetFinalizers()[0] != finalizerName {
 		return nil
 	}
+	var RBAC = mf.Any(mf.ByKind("Role"), mf.ByKind("ClusterRole"), mf.ByKind("RoleBinding"), mf.ByKind("ClusterRoleBinding"))
 	if len(r.servings) == 0 {
 		if err := r.config.Filter(mf.ByKind("Deployment")).Delete(); err != nil {
 			return err
 		}
-		if err := r.config.Filter(mf.NoCRDs).Delete(); err != nil {
+		if err := r.config.Filter(mf.All(mf.NoCRDs, mf.None(RBAC))).Delete(); err != nil {
+			return err
+		}
+		// Delete Roles last, as they may be useful for human operators to clean up.
+		if err := r.config.Filter(RBAC).Delete(); err != nil {
```

Comment:
Why exactly does the order matter? It's not as if they're going to be around long enough for a human to benefit from them are they?

Reply:
In the happy case where the Operator is able to successfully clean up state, yes. In the case where the Operator fails for some reason (user misconfigured RBAC policies, etc) it's better to keep roles like knative-serving-admin ( https://github.com/knative/serving/blob/master/config/core/200-serviceaccount.yaml#L26 ) around so that the human might have the ability to clean things up.

The other reason is https://github.com/knative/serving-operator/pull/291
In that change, the Operator 'bootstrap's itself into having permissions on all resources, by having 'escalate' on all roles, and being automatically granted all clusterroles with 'serving.knative.dev/release', the Operator is able to grant itself any necessary permissions without needing to know exactly what they entail. If the Operator deletes the clusterroles that are allowing it admin access, it may lose the ability to delete resources it still needs to clean up.

Reply:
Gotcha, thanks for the explanation!

Reply:
No problem :) Thanks for the review!